### PR TITLE
add tests for read, write, mkdir, size; make them pass

### PR DESCRIPTION
**Pull request by yomimono**
This PR includes automated tests for mirage-fs-unix's implementations of `read`, `write`, `mkdir`, `connect`, and `size`.  Also included are patches for surprising behaviour revealed by said tests: 

* `connect` now verifies that the argument is a directory, and returns an error if it isn't
* `read` now respects offset and length parameters, rather than returning the entire file
* `write` now respects offset parameter
* `write` no longer opens files with `O_TRUNC`, meaning writes to existing files no longer have the side effect of destroying all pre-existing content
* `write` will return `Error Is_a_directory` when attempting to overwrite a directory
* `size` no longer returns `Ok 4096 for directories
* `mkdir` now returns an error when a directory couldn't be created, and will no longer claim to overwrite files

With respect to the behavior of `write`, I'm not sure this patch is perfect; removing `O_TRUNC` from all writes also unfortunately removes the user's ability to truncate the file at all, short of calling `destroy` on it and then rewriting it in full.  I welcome your thoughts.

### CODE REVIEW

File: .travis-opam.sh [hunk 1]
```diff
@@ -0,0 +1,37 @@
+# If a fork of these scripts is specified, use that GitHub user instead
```
Comment (samoht) [inline .travis-opam.sh]: this file is supposed to be downloaded, so it shouln't be part of the PR

**Comment by samoht** [comment]
Awesome!!! 

Regarding your `write` question, I think we should specify the desired behaviour somewhere (for instance in the `V1.mli`) and ensure that all backend respects the semantics. For instance I'm not sure how the fat implementation currently [behave](https://github.com/mirage/ocaml-fat/blob/master/lib_test/test.ml#L303).

File: lib/FS_unix.ml [hunk 2]
```diff
@@ -117,7 +122,7 @@ let destroy {base} path =
 
 let create {base} path =
   let path = Fs_common.resolve_filename base path in
-  create_directory (Filename.dirname path) >>= fun () ->
+  create_directory (Filename.dirname path) >>= fun _ ->
```
Comment (dsheets) [inline lib/FS_unix.ml]: Should be `function` to match on errors before attempting `touch`?

File: lib/FS_unix.ml [hunk 3]
```diff
@@ -83,20 +75,33 @@ type stat = {
...
+    | `Ok () ->
+      try_lwt
+        Lwt_unix.mkdir path 0o755 >>= fun () -> Lwt.return (`Ok ())
+      with 
+      | Unix.Unix_error (Unix.ENOENT, _, _) -> 
+        Lwt.return (`Error (`No_directory_entry (path, "")))
+      | Unix.Unix_error (Unix.ENOSPC, _, _) ->
+        Lwt.return (`Error (`No_space))
+      | Unix.Unix_error (Unix.EIO, _, _) ->
+        Lwt.return (`Error (`Block_device ()))
```
Comment (dsheets) [inline lib/FS_unix.ml]: Would be nice to break out this error mapping into own function so, e.g., `connect` below can use it.

File: lib/FS_unix.ml [hunk 4]
```diff
@@ -132,6 +137,13 @@ let stat {base} path0 =
   with exn ->
     return (`Error (`No_directory_entry (base, path0)))
 
+let connect id =
+  try_lwt
+    match Sys.is_directory id with
+    | true -> return (`Ok {base = id})
+    | false -> return (`Error (`Not_a_directory id))
+  with (Sys_error _) -> return (`Error (`No_directory_entry (id, "")))
```
Comment (dsheets) [inline lib/FS_unix.ml]: Would be nice to use something like `check_type` from `create_directory` here and then catch the `Unix_error` and use the `create_directory` error map so things like perms errors are handled correctly (right now, will be effectively `ENOENT`).

File: lib/FS_unix.ml [hunk 5]
```diff
@@ -144,21 +156,28 @@ let listdir {base} path =
...
+               Lwt_unix.write fd buf off remaining >>= fun n ->
+               aux (off+n) (remaining-n))
+           in
+           aux 0 (String.length buf) >>= fun () ->
+           return (`Ok ()))
+        (fun e ->
+           Lwt_unix.close fd >>= fun () ->
+           return (`Error (`Unknown_error (Printexc.to_string e))))
+    with
+    | Unix.Unix_error (Unix.EISDIR, _, _) -> return (`Error (`Is_a_directory
```
Comment (dsheets) [inline lib/FS_unix.ml]: Using the error map from above functions would be good.

File: lib/fs_common.ml [hunk 6]
```diff
@@ -59,11 +60,13 @@ let read_impl base name off len =
...
+           return (Some (Cstruct.sub buf 0 (min len reqd_len)))
       )
     in
     Lwt_stream.to_list st >>= fun bufs ->
-    return (`Ok bufs)
+    match reqd_len with
+    | 0 -> return (`Ok [])
+    | n -> return (`Ok bufs)
   with exn ->
     return (`Error (`No_directory_entry (base, name)))
```
Comment (dsheets) [inline lib/fs_common.ml]: Error map here.

File: lib/fs_common.ml [hunk 7]
```diff
@@ -72,7 +75,12 @@ let size_impl base name =
...
-  with exn ->
-    return (`Error (`No_directory_entry (base, name)))
+    match stat.Lwt_unix.LargeFile.st_kind with
+      | Lwt_unix.S_REG -> return (`Ok stat.Lwt_unix.LargeFile.st_size)
+      | _ -> return (`Error (`Is_a_directory name))
+  with 
+    | Unix.Unix_error (Unix.ENOENT, _, _) ->
+      return (`Error (`No_directory_entry (base, name)))
+    | Unix.Unix_error (_, str1, str2) -> 
+      return (`Error (`Unknown_error (str1 ^ str2)))
```
Comment (dsheets) [inline lib/fs_common.ml]: And error map here.

**Comment by dsheets** [comment]
I wonder about adding a function like the `truncate` syscall. Also, do we have a solution for `fsync`/`fdatasync`?

**Comment by mor1** [comment]
(Idle thought -- could this API and/or implementations make use of the
submitted work on POSIX FS semantics?)

On 13 May 2015 at 12:15, David Sheets <notifications@github.com> wrote:

> I wonder about adding a function like the truncate syscall. Also, do we
> have a solution for fsync/fdatasync?
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/mirage/mirage-fs-unix/pull/10#issuecomment-101610382>.
>



-- 
Richard Mortier
mort@cantab.net

File: lib/FS_unix.ml (hunk 3 above)
Comment (yomimono) [inline lib/FS_unix.ml]: OK, I'll add that and references to it as you've mentioned in the other line notes.  Thanks!

**Comment by yomimono** [comment]
@dsheets - no sync invocations at all, as far as I can see; the implementation seems to assume they're not required.

**Comment by dsheets** [comment]
Unfortunately, I think this API is exactly the opposite of the existing semantics work in that it is a consumer of the POSIX API and not a producer. If we wrote down the semantics of this API in Lem or a theorem prover and then wrote this layer in Lem, we could prove that the Mirage semantics are obeyed by this layer when supplied with a POSIX FS.

This API would be a good test case for a "simple" consumer of a POSIX API. Perhaps we could do something with the spec and recorded traces of this API? Or maybe build some kind of nondeterministic test harness which could generate POSIX behaviors that this API doesn't cover?

I believe a significant amount of work would be required to leverage the POSIX spec here but it might be worth it with a good approach.

**Comment by yomimono** [comment]
@samoht - It looks to me like ocaml-fat doesn't destroy existing files on write: https://github.com/yomimono/ocaml-fat/blob/test_write_truncate/lib_test/test.ml#L332

**Comment by samoht** [comment]
I'm fine to merge theses even with the semantic merge for `write`. The only consumer of the library seems to be `mirage-www` which do not use the write part anyway.

At one point in the future we can imagine having base tests being functors taking `V1.KV_RO` and `V1.FS` as argument (and distributed with `mirage-types`) and testing the correctness of all the implementations.

**Comment by mor1** [comment]
Perhaps worth taking this to the MirageOS devel / REMS list(s)?
Seems like a longer term, research oriented thing that may be of interest...

On 13 May 2015 at 11:37, David Sheets <notifications@github.com> wrote:

> Unfortunately, I think this API is exactly the opposite of the existing
> semantics work in that it is a consumer of the POSIX API and not a
> producer. If we wrote down the semantics of this API in Lem or a theorem
> prover and then wrote this layer in Lem, we could prove that the Mirage
> semantics are obeyed by this layer when supplied with a POSIX FS.
>
> This API would be a good test case for a "simple" consumer of a POSIX API.
> Perhaps we could do something with the spec and recorded traces of this
> API? Or maybe build some kind of nondeterministic test harness which could
> generate POSIX behaviors that this API doesn't cover?
>
> I believe a significant amount of work would be required to leverage the
> POSIX spec here but it might be worth it with a good approach.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/mirage/mirage-fs-unix/pull/10#issuecomment-101619243>.
>



-- 
Richard Mortier
mort@cantab.net

**Pull request merged**